From a4d4ce4a5e3ffe484b13328cf9bca188910607b6 Mon Sep 17 00:00:00 2001
From: Zhou Guanghui <zhouguanghui1@huawei.com>
Date: Thu, 9 Sep 2021 07:54:19 +0000
Subject: [PATCH] driver/svm: support SVSP feature, interface definition

ascend inclusion
category: bugfix
bugzilla: NA
DTS: NA
CVE: NA

-----------------------------------------------------------------

In the SVA scenario, the master device of the SMMU can access the
address space of the process by the operation of binding proces.
The page table of the process is shared by the MMU and SMMU, In
this case, if some addrss range cannot be accessed through the MMU,
access to this address range needs to be isolated from the process
page table.

If the default domain of the SMMU is used, the isolation can be
achieved. However, the address space created by default domain do
not like the address space of the process which can be shared by
multiple SMMUs.

Therefore, we add a shadow virtual space of process. This address
space is used only by the SMMU.

Signed-off-by: Zhou Guanghui <zhouguanghui1@huawei.com>
---
 arch/arm64/Kconfig          | 15 +++++++++++++++
 include/linux/ascend_smmu.h | 31 +++++++++++++++++++++++++++++++
 kernel/fork.c               |  1 +
 mm/mmap.c                   |  1 +
 4 files changed, 48 insertions(+)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 3369438a5..e95d8a06c 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -2156,6 +2156,21 @@ config ASCEND_SHARE_POOL
 	  This feature allows multiple processes to share virtual memory both
 	  in kernel and user level, which is only enabled for ascend platform.
 
+config ASCEND_SVSP
+	bool "Enable support for SVSP"
+	default n
+	depends on ARM_SMMU_V3_SVA
+	help
+	  SVSP means Shadow Virtual Space of Process, which is an IO space used for
+	  master device of SMMU.
+
+	  This space is only used for the master device, since the target address
+	  after traslation by MMU is not an accessible physical address, but can
+	  accessible by SMMU.
+
+	  Therefore, in the sva scenario, an svsp is created for each target process
+	  address space.
+
 endif
 
 endmenu
diff --git a/include/linux/ascend_smmu.h b/include/linux/ascend_smmu.h
index 98bc72361..d7b5c5fdf 100644
--- a/include/linux/ascend_smmu.h
+++ b/include/linux/ascend_smmu.h
@@ -8,6 +8,37 @@ extern int ascend_smmu_get_mpam(int pasid, int *partid, int *pmg, int *s1mpam);
 extern int ascend_smmu_set_user_mpam_en(int user_mpam_en);
 extern int ascend_smmu_get_user_mpam_en(int *user_mpam_en);
 
+extern int svm_get_pasid(pid_t vpid, int dev_id);
+
+#ifdef CONFIG_ASCEND_SVSP
+extern unsigned long svm_svsp_mmap(unsigned long len, int pasid);
+extern int svm_svsp_remap_range(unsigned long va, unsigned long iova,
+		size_t size, int pasid, pgprot_t prot);
+extern void svm_svsp_munmap(unsigned long start, unsigned long len, int pasid);
+extern struct mm_struct *svm_svsp_of_mm(struct mm_struct *mm);
+#else
+static inline unsigned long svm_svsp_mmap(unsigned long len, int pasid)
+{
+	return -EINVAL;
+}
+
+static inline int svm_svsp_remap_range(unsigned long va, unsigned long iova,
+		size_t size, int pasid, pgprot_t prot)
+{
+	return -EINVAL;
+}
+
+static inline void svm_svsp_munmap(unsigned long start, unsigned long len, int pasid)
+{
+	return;
+}
+
+static inline struct mm_struct *svm_svsp_of_mm(struct mm_struct *mm)
+{
+	return NULL;
+}
+#endif
+
 #ifndef __LINUX_AGENT_SMMU_H
 #define __LINUX_AGENT_SMMU_H
 
diff --git a/kernel/fork.c b/kernel/fork.c
index a2f84e007..5ed5b85a9 100644
--- a/kernel/fork.c
+++ b/kernel/fork.c
@@ -1129,6 +1129,7 @@ struct mm_struct *mm_alloc(void)
 
 	return mm_init(mm, current, current_user_ns());
 }
+EXPORT_SYMBOL_GPL(mm_alloc);
 
 static inline void __mmput(struct mm_struct *mm)
 {
diff --git a/mm/mmap.c b/mm/mmap.c
index bddd7f0f8..4a2bf19ef 100644
--- a/mm/mmap.c
+++ b/mm/mmap.c
@@ -3024,6 +3024,7 @@ int __do_munmap(struct mm_struct *mm, unsigned long start, size_t len,
 
 	return downgrade ? 1 : 0;
 }
+EXPORT_SYMBOL(__do_munmap);
 
 int do_munmap(struct mm_struct *mm, unsigned long start, size_t len,
 	      struct list_head *uf)
-- 
2.17.1

